include/mockturtle/algorithms/mig_inv_optimization.hpp

```diff
@@ -0,0 +1,355 @@
+    if ( a.index == b.index )
+    {
+      return ( a.complement == b.complement ) ? a : c;
+    }
+    else if ( b.index == c.index )
+    {
+      return ( b.complement == c.complement ) ? b : a;
+    }
+
+    std::shared_ptr<storage<regular_node<3, 2, 1>>>::element_type::node_type nd;
+    nd.children[0] = a;
+    nd.children[1] = b;
+    nd.children[2] = c;
+
+    /* structural hashing */
+    const auto it = ntk._storage->hash.find( nd );
+    if ( it != ntk._storage->hash.end() )
+    {
+      return { it->second, 0 };
+    }
+
+    const auto index = ntk._storage->nodes.size();
+
+    if ( index >= .9 * ntk._storage->nodes.capacity() )
+    {
+      ntk._storage->nodes.reserve( static_cast<uint64_t>( 3.1415f * index ) );
+      ntk._storage->hash.reserve( static_cast<uint64_t>( 3.1415f * index ) );
+    }
+
+    ntk._storage->nodes.push_back( nd );
+
+    ntk._storage->hash[nd] = index;
+
+    /* increase ref-count to children */
+    ntk._storage->nodes[a.index].data[0].h1++;
+    ntk._storage->nodes[b.index].data[0].h1++;
+    ntk._storage->nodes[c.index].data[0].h1++;
+
+    for ( auto const& fn : ntk._events->on_add )
+    {
+      ( *fn )( index );
+    }
+
+    return { index, 0 };
+  }
+
+private:
+  Ntk& ntk;
+  mig_inv_optimization_stats& st;
+};
+
+} // namespace detail
+
+/*! \brief MIG inverter optimization.
+ *
+ * This algorithm tries to reduce the number
+ * of inverters in a MIG network without
+ * increasing the node number. It checks each
+ * node for 1 level and 2 level optimization
+ * opportunuties and inverts the node if it
+ * decreases the number of inverted connections.
+ * It does not count constant values as inverted
+ * even if they are complemented in the graph.
+ *
+ * **Required network functions:**
+ * 'foreach_fanin'
+ * 'foreach_fanout'
+ * 'substitute_node'
+ * 'replace_in_outputs'
+ * 'is_complemented'
+ * 'get_node'
+ * 'is_dead'
+ * 'is_constant'
+ * 'foreach_gate'
+ */
+template<class Ntk>
+void mig_inv_optimization( Ntk& ntk, mig_inv_optimization_stats* pst = nullptr )
+{
+  static_assert( is_network_type_v<Ntk>, "Ntk is not a network type" );
+  static_assert( has_foreach_fanin_v<Ntk>, "Ntk does not implement the foreach_fanin method" );
+  static_assert( has_foreach_fanout_v<Ntk>, "Ntk does not implement the foreach_fanout method" );
+  static_assert( has_substitute_node_v<Ntk>, "Ntk does not implement the substitute_node method" );
+  static_assert( has_replace_in_outputs_v<Ntk>, "Ntk does not implement the replace_in_outputs method" );
+  static_assert( has_is_complemented_v<Ntk>, "Ntk does not implement the is_complemented method" );
+  static_assert( has_get_node_v<Ntk>, "Ntk does not implement the get_node method" );
+  static_assert( has_is_dead_v<Ntk>, "Ntk does not implement the is_dead method" );
+  static_assert( has_is_constant_v<Ntk>, "Ntk does not implement the is_constant method" );
+  static_assert( has_foreach_gate_v<Ntk>, "Ntk does not implement the foreach_gate method" );
```

Comment:
Shouldn't you add a `static_assert` check for the MIG base type? Based on the code and the original paper, it seems like the algorithm is only applicable to MIGs.